[macros] Introduce \indexhdr and apply it for all headers.

jensmaurer opened the pull request:
Partially addresses #1807.

jensmaurer commented:
This was sed-converted from the status quo. We can redefine \indexhdr to something else at a later time.

zygoloid commented:
This will need some more changes once we land the motions (we're adding two new headers), but this looks good, thanks!

jensmaurer commented:
I've rebased this, fixed mentions of the additional headers, and while doing so, fixed bad index entries <<syncstream>> and <<type_traits>>. (\idxhdr implicitly adds the <>.)
